[PTD][TP] Add BWD support for colwise embedding sharding

fduwjj opened the pull request:
Originally, we didn't enable BWD for colwise embedding because we thought it was just for inference, but it turns out that we do need it for training. So, let's enable it for now and unit test is also added.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104820

fduwjj committed:
Add BWD for colwise embedding table sharding

[ghstack-poisoned]

fduwjj committed:
Update on "[PTD][TP] Add BWD support for colwise embedding sharding"

Originally, we didn't enable BWD for colwise embedding because we thought it was just for inference, but it turns out that we do need it for training. So, let's enable it for now and unit test is also added.




[ghstack-poisoned]

fegin reviewed (approved):
LGTM, but can you check if the failing distributed tests are related to this PR?

fduwjj commented:
@pytorchbot rebase

fduwjj committed:
Update on "[PTD][TP] Add BWD support for colwise embedding sharding"

Originally, we didn't enable BWD for colwise embedding because we thought it was just for inference, but it turns out that we do need it for training. So, let's enable it for now and unit test is also added.




[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge

fduwjj commented:
@pytorchbot merge
